go-controller/pkg/ovn/namespace.go

```diff
@@ -135,112 +134,137 @@ func (oc *Controller) multicastUpdateNamespace(ns *kapi.Namespace) {
 
 	// Get all the pods in the namespace and append their IP to the
 	// address_set
 	existingPods, err := oc.watchFactory.GetPods(ns.Name)
+	addresses := make([]string, 0)
 	if err != nil {
 		klog.Errorf("Failed to get all the pods (%v)", err)
 	} else {
 		for _, pod := range existingPods {
 			if pod.Status.PodIP != "" {
 				portName := podLogicalPortName(pod)
-				oc.namespaceAddressSet[ns.Name][pod.Status.PodIP] = portName
+				nsInfo.addressSet[pod.Status.PodIP] = portName
+				addresses = append(addresses, pod.Status.PodIP)
 			}
 		}
 	}
 
-	addresses := make([]string, 0)
-	for address := range oc.namespaceAddressSet[ns.Name] {
-		addresses = append(addresses, address)
-	}
-
 	// Create an address_set for the namespace.  All the pods' IP address
 	// in the namespace will be added to the address_set
 	createAddressSet(ns.Name, hashedAddressSet(ns.Name), addresses)
 
-	oc.namespacePolicies[ns.Name] = make(map[string]*namespacePolicy)
-	oc.multicastUpdateNamespace(ns)
+	oc.multicastUpdateNamespace(ns, nsInfo)
 }
 
 func (oc *Controller) updateNamespace(old, newer *kapi.Namespace) {
-	klog.V(5).Infof("Updating namespace: old %s new %s", old.Name, newer.Name)
+	klog.V(5).Infof("Updating namespace: %s", old.Name)
 
-	// A big fat lock per namespace to prevent race conditions
-	// with namespace resources like address sets and deny acls.
-	oc.namespaceMutex[newer.Name].Lock()
-	defer oc.namespaceMutex[newer.Name].Unlock()
+	nsInfo := oc.getNamespaceLocked(old.Name)
+	if nsInfo == nil {
+		klog.Warningf("Update event for unknown namespace %q", old.Name)
+		return
+	}
+	defer nsInfo.Unlock()
 
-	oc.multicastUpdateNamespace(newer)
+	oc.multicastUpdateNamespace(newer, nsInfo)
 }
 
 func (oc *Controller) deleteNamespace(ns *kapi.Namespace) {
 	klog.V(5).Infof("Deleting namespace: %s", ns.Name)
-	oc.namespaceMutexMutex.Lock()
-	defer oc.namespaceMutexMutex.Unlock()
 
-	mutex, ok := oc.namespaceMutex[ns.Name]
-	if !ok {
+	nsInfo := oc.deleteNamespaceLocked(ns.Name)
+	if nsInfo == nil {
 		return
 	}
-	mutex.Lock()
-	defer mutex.Unlock()
+	defer nsInfo.Unlock()
 
 	deleteAddressSet(hashedAddressSet(ns.Name))
-	oc.multicastDeleteNamespace(ns)
-	delete(oc.namespacePolicies, ns.Name)
-	delete(oc.namespaceAddressSet, ns.Name)
-	delete(oc.namespaceMutex, ns.Name)
+	oc.multicastDeleteNamespace(ns, nsInfo)
 }
 
-// getNamespaceLock grabs the lock for a particular namespace. If the
-// namespace does not exist, returns nil. Otherwise, returns the held lock.
-func (oc *Controller) getNamespaceLock(ns string) *sync.Mutex {
-	// lock the list of namespaces, get the mutex
-	oc.namespaceMutexMutex.Lock()
-	mutex, ok := oc.namespaceMutex[ns]
-	oc.namespaceMutexMutex.Unlock()
-	if !ok {
+// getNamespaceLocked locks namespacesMutex, looks up ns, and (if found), returns it with
+// its mutex locked. If ns is not known, nil will be returned
+func (oc *Controller) getNamespaceLocked(ns string) *namespaceInfo {
+	oc.namespacesMutex.Lock()
+	nsInfo := oc.namespaces[ns]
```

Comment:
is it ever possible that the map entry does not exist? should you check to be safe?

Reply:
If the map entry doesn't exist then we'll get back `nil` and we do check that.
(the `, ok` form is only needed if some elements of the map are explicitly set to the zero value and you need to distinguish that from them not being set)